src/random_events.rs

```diff
@@ -36,13 +42,29 @@ impl RandomEvents {
         Event::AddNode(random())
     }
 
-    fn random_remove(&self, nodes: &BTreeMap<Name, Node>) -> Event {
-        Event::RemoveNode(Self::find_node_to_remove(nodes))
+    fn random_remove(&self, nodes: &BTreeMap<Name, Node>) -> Option<Event> {
+        self.find_node_to_remove(nodes).map(Event::RemoveNode)
     }
 
-    fn find_node_to_remove(nodes: &BTreeMap<Name, Node>) -> Name {
-        sample_single(nodes.iter())
-            .map(|(name, _)| *name)
-            .unwrap()
+    // Remove a randomly-selected node which is in a section with at least quorum + 1 members. The
+    // section's member count is calculated by removing any dead nodes from the node's own current
+    // block's member list. If no suitable node can be found, the function returns `None`.
+    fn find_node_to_remove(&self, nodes: &BTreeMap<Name, Node>) -> Option<Name> {
+        let names_sorted: BTreeSet<_> = nodes.keys().cloned().collect();
+        let mut names = nodes.keys().cloned().collect_vec();
+        shuffle(&mut names);
+        for name in names {
+            if let Some(our_current_block) = nodes[&name].our_current_blocks().next() {
+                let section_size = our_current_block
+                    .members
+                    .intersection(&names_sorted)
+                    .count();
+                if (section_size - 1) * 2 > self.node_params.min_section_size {
```

Comment:
I think `min_section_size` should be replaced by `our_current_block.members.len()` here. In the case of the faulty seed you linked on Slack (`[1121350019, 3611361785, 669310913, 2290061490]`), the section ends up with a block that has 13 members, of which only 5 are still alive! 5 is a quorum of min_section_size 8, but not of the actual section size of 13, so the network is stalled at this point.

I've got some code (that I'll push somewhere) that checks that all the nodes present in current blocks *are* actually alive. I think we need this check, or something similar.